azurerm/resource_arm_storage_account_blob_settings.go

```diff
@@ -0,0 +1,177 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+var storageAccountBlobSettingsResourceName = "azurerm_storage_account_blob_settings"
+
+func resourceArmStorageAccountBlobSettings() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmStorageAccountBlobSettingsCreateUpdate,
+		Read:   resourceArmStorageAccountBlobSettingsRead,
+		Update: resourceArmStorageAccountBlobSettingsCreateUpdate,
+		Delete: resourceArmStorageAccountBlobSettingsDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"storage_account_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"enable_soft_delete": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"soft_delete_retention_days": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Default:      7,
+				ValidateFunc: validation.IntBetween(1, 365),
+			},
```

Comment:
since the `azurerm_storage_account` resource already supports configuring `queue_properties`, since you can only have one of these would this make more sense within that resource e.g.

```
resource "azurerm_storage_account" "test" {
  ...

  blob_properties {
    soft_deletion_policy {
      days = 7
    }
  }
}
```

we could then assume the `soft_deletion_policy` block being omited/empty wants this disabled - which means that users wouldn't need to set `enabled` or not - WDYT?

from Terraform's perspective that'd create an Expand / Flatten function like so:

```
func expandStorageAccountBlobProperties(input []interface{}) storage.BlobServiceProperties {
	properties := storage.BlobServiceProperties{
		BlobServicePropertiesProperties: &storage.BlobServicePropertiesProperties{
			DeleteRetentionPolicy: &storage.DeleteRetentionPolicy{
				Enabled: utils.Bool(false),
			},
		},
	}
	if len(input) == 0 {
		return properties
	}

	raw := input[0].(map[string]interface{})
	deletionPolicyRaw := raw["delete_retention_policy"].([]interface{})
	if len(deletionPolicyRaw) > 0 {
		policyRaw := deletionPolicyRaw[0].(map[string]interface{})
		days := policyRaw["days"].(int)
		properties.BlobServicePropertiesProperties.DeleteRetentionPolicy.Enabled = utils.Bool(true)
		properties.BlobServicePropertiesProperties.DeleteRetentionPolicy.Days = utils.Int32(int32(days))
	}

	return properties
}

func flattenStorageAccountBlobProperties(input *storage.BlobServiceProperties) []interface{} {
	enabled := false
	days := 0

	if input != nil && input.BlobServicePropertiesProperties != nil && input.BlobServicePropertiesProperties.DeleteRetentionPolicy != nil {
		if policy := input.BlobServicePropertiesProperties.DeleteRetentionPolicy; policy != nil {
			if policy.Enabled != nil {
				enabled = *policy.Enabled
			}

			if policy.Days != nil {
				days = int(*policy.Days)
			}
		}
	}

	output := make([]interface{}, 0)
	if enabled {
		output = append(output, map[string]interface{}{
			"days": days,
		})
	}

	return []interface{}{
		map[string]interface{}{
			"delete_retention_policy": output,
		},
	}
}
```

WDYT?

Reply:
Sounds fine for me, I'll change that!

Reply:
done